src/array.cr

```diff
@@ -1133,7 +1133,7 @@ class Array(T)
   # ```
   #
   # See also: `Array#select`.
-  def select!
+  def select!(& : T ->) : self
```

Comment:
```suggestion
  def select!(& : T -> Bool) : self
```

Is there a reason this and like `reject(!)` shouldn't be forced to return a `Bool`?

Reply:
That would definitely be a breaking change. Right now, the value just nees to be interpreted as truthy. And I think that's probably good. No need to force users to a bool conversion.

Reply:
Wouldn't `T ->` mean `Nil` return value type?

Reply:
Nope. It means there's no restriction on the return type. Basically like a free var.

Reply:
I think @Sija is right. You need to use underscore to mean any type

Reply:
Actually, this just matters for captured blocks

Reply:
It's equivalent to `T -> _`, which means "allow any return type".

Reply:
Yes, it's different for captured blocks.

/cc https://forum.crystal-lang.org/t/difference-between-underscore-and-empty-as-block-return-value/2998

Reply:
Ideally we should probably improve the semantics, maybe make empty invalid for non-captured blocks and require an underscore since that's more explicit and fits with the use of underscore in the type grammar.

But for now, it's better to use empty because it's not entirely equivalent to underscore: https://forum.crystal-lang.org/t/difference-between-underscore-and-empty-as-block-return-value/2998

Reply:
See #10932 for the underscore issue and 10931 for a discussion about the intended semantics.

Reply:
Based on #11210 we never constrain the return type to `Bool` in contexts like this.